### PR TITLE
Feature/transaction

### PR DESCRIPTION
**Pull request by Aleksey-Kn**

### CODE REVIEW

File: src/main/java/io/github/alekseykn/imnorm/FastRepository.java [hunk 1]
```diff
@@ -50,36 +50,50 @@ protected synchronized Record create(String id, Record record) {
...
     }
 
     @Override
-    public Set<Record> findAll() {
-        waitAllRecords();
-        synchronized (this) {
-            return data.values().stream().map(Cluster::findAll).flatMap(Collection::stream).collect(Collectors.toSet());
+    protected synchronized Record create(String id, Record record, Transaction transaction) {
+        if (needGenerateId) {
+            id = generateAndSetIdForRecord(record);
```
Comment (EnKarin) [inline src/main/java/io/github/alekseykn/imnorm/FastRepository.java:69]: It is not necessary to change the input parameters

File: src/main/java/io/github/alekseykn/imnorm/FastRepository.java [hunk 2]
```diff
@@ -50,36 +50,50 @@ protected synchronized Record create(String id, Record record) {
...
+            currentCluster.set(id, record, transaction);
         }
+        return record;
     }
 
     @Override
-    public Set<Record> findAll(int startIndex, int rowCount) {
-        HashSet<Record> result = new HashSet<>(rowCount);
+    public synchronized Set<Record> findAll() {
+        return data.values().stream().map(Cluster::findAll).flatMap(Collection::stream).collect(Collectors.toSet());
```
Comment (EnKarin) [inline src/main/java/io/github/alekseykn/imnorm/FastRepository.java:83]: you need to divide stream line by line

File: src/main/java/io/github/alekseykn/imnorm/FastRepository.java [hunk 3]
```diff
@@ -101,46 +115,71 @@ public Set<Record> findAll(int startIndex, int rowCount) {
...
-                } else if (pastFirstKey.equals(realId)) {
-                    Files.delete(Path.of(directory.getAbsolutePath(), realId));
-                    data.remove(realId);
-                    data.put(cluster.firstKey(), cluster);
-                }
-            } catch (IOException e) {
-                throw new InternalImnormException(e);
-            }
-            return record;
+            clustersData = data.values().stream().map(Cluster::findAll).collect(Collectors.toList());
```
Comment (EnKarin) [inline src/main/java/io/github/alekseykn/imnorm/FastRepository.java:121]: similarly

File: src/main/java/io/github/alekseykn/imnorm/FrugalRepository.java [hunk 4]
```diff
@@ -75,110 +75,189 @@ protected synchronized Record create(String id, Record record) {
...
-                            return Files.lines(Path.of(directory.getAbsolutePath(), clusterName));
-                        } catch (IOException e) {
-                            throw new InternalImnormException(e);
-                        }
-                    })
-                    .map(record -> gson.fromJson(record, type))
-                    .collect(Collectors.toSet());
+    protected synchronized Record create(String id, Record record, Transaction transaction) {
+        if (needGenerateId) {
+            id = generateAndSetIdForRecord(record);
```
Comment (EnKarin) [inline src/main/java/io/github/alekseykn/imnorm/FrugalRepository.java:80]: no need to change the input parameters

Comment (Aleksey-Kn) [reply src/main/java/io/github/alekseykn/imnorm/FrugalRepository.java]: Taken out to Repository.save too

File: src/main/java/io/github/alekseykn/imnorm/FastRepository.java (hunk 3 above)
Comment (Aleksey-Kn) [reply src/main/java/io/github/alekseykn/imnorm/FastRepository.java]: Done

File: src/main/java/io/github/alekseykn/imnorm/FastRepository.java (hunk 2 above)
Comment (Aleksey-Kn) [reply src/main/java/io/github/alekseykn/imnorm/FastRepository.java]: Done

**Review by EnKarin — APPROVED** [review]

**Pull request merged**